src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs

```diff
@@ -141,54 +141,98 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
+            }
+            catch (Exception ex)
+            {
+                Logger.Error($"Failed to retreive headers from poison message '{messageId}'. Moving message to queue '{settings.ErrorQueue}'...", ex);
+                await Move(message, settings.ErrorQueue, messageId).ConfigureAwait(false);
+
+                return;
+            }
+
+            using (var tokenSource = new CancellationTokenSource())
+            {
+                var pushContext = new PushContext(messageId, headers, new MemoryStream(message.Body ?? new byte[0]), new TransportTransaction(), tokenSource, new ContextBag());
+                await pipe(pushContext).ConfigureAwait(false);
 
-                try
+                if (tokenSource.IsCancellationRequested)
                 {
-                    messageId = messageConverter.RetrieveMessageId(message);
-                    headers = messageConverter.RetrieveHeaders(message);
+                    await ReturnToQueue(message.DeliveryTag).ConfigureAwait(false);
                 }
-                catch (Exception ex)
+                else
                 {
-                    await poisonMessageForwarder.ForwardPoisonMessageToErrorQueue(message, ex, settings.ErrorQueue).ConfigureAwait(false);
                     await Acknowledge(message.DeliveryTag, messageId).ConfigureAwait(false);
-
-                    return;
                 }
+            }
+        }
 
-                using (var tokenSource = new CancellationTokenSource())
-                {
-                    var pushContext = new PushContext(messageId, headers, new MemoryStream(message.Body ?? new byte[0]), new TransportTransaction(), tokenSource, new ContextBag());
-                    await pipe(pushContext).ConfigureAwait(false);
-
-                    if (tokenSource.IsCancellationRequested)
-                    {
-                        await Requeue(message.DeliveryTag, messageId).ConfigureAwait(false);
-                    }
-                    else
-                    {
-                        await Acknowledge(message.DeliveryTag, messageId).ConfigureAwait(false);
-                    }
-                }
+        async Task Move(BasicDeliverEventArgs message, string queue, string messageId = null)
+        {
+            try
+            {
+                await poisonMessageForwarder.ForwardPoisonMessageToErrorQueue(message, settings.ErrorQueue).ConfigureAwait(false);
             }
             catch (Exception ex)
             {
-                Logger.Warn($"Error while attempting to process message {messageId}. The message will be requeued.", ex);
-                await Requeue(message.DeliveryTag, messageId).ConfigureAwait(false);
+                Logger.Error($"Failed to move message{(messageId == null ? "" : " '" + messageId + "'")} to queue '{queue}'. Returning message to original queue...", ex);
+                await ReturnToQueue(message.DeliveryTag).ConfigureAwait(false);
```

Comment:
Are you sure this is the right behavior? This could lead to indefinite retries. Would it make more sense to trigger critical error or return to queue and arm the circuit breaker?